contributor-docs/adrs/adr-016-css.md

```diff
@@ -0,0 +1,103 @@
+# ADR 16: Styling with CSS
+
+## Status
+
+| Stage    | Status                        |
+| -------- | ----------------------------- |
+| Approved | ✅                            |
+| Adopted  | 🚧 _TODO: add tracking issue_ |
+
+## Context
+
+Reasons for change: Performance, utilise css variables
+
+The main reason for changing our styling architecture is to remove performance issues for applications:
+
+Relevant parts from [primer/discussions#1928](https://github.com/github/primer/discussions/1928#reasons):
+
+1. Initial page loads take longer
+
+   Because styles are injected during runtime, we can not parallelize this work. The page is un-styled till the javascript bundle is
+   downloaded and processed. The processing time is longer on slower devices.
+
+   Benchmarks:
+
+   i. Lab benchmarks: rendering a page with 1000 components
+   runtime injection of css: 242ms
+   static .css files: 96ms (60% faster)
+
+   ii. Application benchmarks: render times for repo directory (>1000 files, [gh/gh#241090](https://github.com/github/github/pull/241090))
+   with primer/react Box: 3800ms
+   with custom css: 3076ms (20% faster)
+
+2. Slow server side rendering
+
+   Some applications are able to make up for slower initial render times by collecting styles on the server first. In the correct stack, collecting styles requires rendering the app twice before we can ship any html to the user.
+
+   benchmarks: in the issues-app experiment, collecting styles on the server contributed to + 20% of rendering time (450ms on top of 1900ms)
+
+3. Style updates are expensive
+
+   The slowest part of the current stack is updating styles based on props or state. Big changes take linearly longer, worse on slower devices.
+
+   Benchmarks:
+
+   i. Lab benchmarks: updating a page with 1000 components, static runtime css vs. dynamic runtime css (thanks @mattcosta7 for these!)
+
+   re-rendering components with fixed styles on runtime: 242ms
+   re-rendering components with dynamic styles on runtime: 441ms (55% slower)
+
+   ii. Application benchmarks: opting out of IconButton for PR diff view (>500 files, [gh/gh#243269](https://github.com/github/github/pull/243269))
+
+   dynamic styles with sx prop: 400ms
+   dynamic styles with .css files: 165ms (60% faster)
+
+The other reasons for change are to utilise css variables for theming and improve the guardrails of consuming primer/primitives
+
+## Decisions
+
+1. Manually refactor styles to css files + css modules
```

Comment:
Would the scope of this ADR include techniques for:

- Sharing between CSS files
- Managing dependencies between CSS files

Here are some situations that I was thinking of:

- I have a helper class that I want to make sure is available to a component (like `sr-only`)
- I have a helper mixin (in the sass sense) to reset a button in order to customize styling
- I have a helper mixin or function to help with managing breakpoints

Reply:
Haven't included any of these. Hoping to define those as we go.

> ### Out of scope
> 
> Methodologies and guidelines on how to author CSS are not part of this ADR/decision. We will define those as we move components to css and find repeating patterns.

https://github.com/primer/react/blob/adr-css/contributor-docs/adrs/adr-016-css.md#out-of-scope

Reply:
@siddharthkp these could be worth bringing up at this point as it would influence the technical decision (in this case CSS Modules) as the different approaches have pros/cons when it comes to share and re-use.

Or should I read this ADR as more about styling with CSS (as opposed to styled-components) and that decisions like CSS Modules are not a part of the scope of this ADR?

Reply:
> Or should I read this ADR as more about styling with CSS (as opposed to styled-components) and that decisions like CSS Modules are not a part of the scope of this ADR?

That was my intention, but now I think it makes sense to include css modules as part of the decision